File: orca/whirlpool/python_client/essentials_alpha/whirlpool_essentials/swaputil.py
from typing import List, NamedTuple
from solana.publickey import PublicKey
from .constants import U64_MAX, MIN_SQRT_PRICE, MAX_SQRT_PRICE, MAX_SWAP_TICK_ARRAYS
from .tickutil import TickUtil
from .pdautil import PDAUtil


class SwapUtil:
    # https://orca-so.github.io/whirlpools/classes/SwapUtils.html#getTickArrayPublicKeys
    # https://github.com/orca-so/whirlpools/blob/2df89bb/sdk/src/utils/public/swap-utils.ts#L75
    @staticmethod
    def get_tick_array_pubkeys(
        tick_current_index: int,
        tick_spacing: int,
        a_to_b: bool,
        program_id: PublicKey,
        whirlpool_pubkey: PublicKey,
    ) -> List[PublicKey]:
        offset = 0
        pubkeys = []
        for i in range(MAX_SWAP_TICK_ARRAYS):
            start_tick_index = TickUtil.get_start_tick_index(tick_current_index, tick_spacing, offset)
            tick_array_pubkey = PDAUtil.get_tick_array(program_id, whirlpool_pubkey, start_tick_index).pubkey
            pubkeys.append(tick_array_pubkey)
            offset = (offset - 1) if a_to_b else (offset + 1)
        return pubkeys

    # https://orca-so.github.io/whirlpools/classes/SwapUtils.html#getDefaultSqrtPriceLimit
    # https://github.com/orca-so/whirlpools/blob/2df89bb/sdk/src/utils/public/swap-utils.ts#L28
    @staticmethod
    def get_default_sqrt_price_limit(a_to_b: bool) -> int:
        return MIN_SQRT_PRICE if a_to_b else MAX_SQRT_PRICE

    # https://orca-so.github.io/whirlpools/classes/SwapUtils.html#getDefaultOtherAmountThreshold
    # https://github.com/orca-so/whirlpools/blob/7b9ec35/sdk/src/utils/public/swap-utils.ts#L37
    @staticmethod
    def get_default_other_amount_threshold(amount_specified_is_input: bool) -> int:
        return 0 if amount_specified_is_input else U64_MAX

File: orca/pool/rust_client/with_anchor_client/src/whirlpool_essential/swap_util.rs
use anchor_client::solana_sdk::pubkey::Pubkey;
use whirlpool::math::tick_math::{MAX_SQRT_PRICE_X64, MIN_SQRT_PRICE_X64};

use super::{pda_util, tick_util};

pub fn get_tick_array_pubkeys(
  tick_current_index: i32,
  tick_spacing: u16,
  a_to_b: bool,
  program_id: &Pubkey,
  whirlpool_pubkey: &Pubkey,
) -> [Pubkey; 3] {
  let mut offset = 0;
  let mut pubkeys: [Pubkey; 3] = Default::default();
  let shifted = if a_to_b { 0i32 } else { tick_spacing as i32 };

  for i in 0..pubkeys.len() {
      let start_tick_index = tick_util::get_start_tick_index(tick_current_index + shifted, tick_spacing, offset);
      let tick_array_pubkey = pda_util::get_tick_array(program_id, whirlpool_pubkey, start_tick_index).pubkey;
      pubkeys[i] = tick_array_pubkey;
      offset = if a_to_b { offset - 1 } else { offset + 1 };
  }

  pubkeys
}

pub fn get_default_sqrt_price_limit(a_to_b: bool) -> u128 {
  if a_to_b {
      MIN_SQRT_PRICE_X64
  } else {
      MAX_SQRT_PRICE_X64
  }
}
